library/core/tests/iter.rs

```diff
@@ -3509,7 +3509,7 @@ pub fn extend_for_unit() {
 #[test]
 fn test_intersperse() {
     let xs = ["a", "", "b", "c"];
-    let v: Vec<&str> = xs.iter().map(|x| x.clone()).intersperse(", ").collect();
```

Comment:
@Aaron1011 We've already talked about this, but I am worried about calling this call a "noop" might make some users confused. The method call has the effect of dereferencing the double reference. The specific method itself does nothing but the mere act of making a method call _does_ have an effect. This distinction is subtle, and I'm worried this may lead to situations like this one where we recommend removing the method call which actually breaks the code.

cc @estebank

Reply:
I think that what _could_ be done is check the type of `expr` and see if it is different to the type of the method receiver (check if one is and the other is not a borrow). The "correct" check would be to see if this is coming from a `impl Clone &T`, but that seems like it would be harder to implement than the proposed heuristic.

Reply:
i tried to do this, but I was not ever able to resolve the return type of the method to something that could be equated to the receiver. I'm not sure how to resolve the return type of the method (which is `Self`) to a type that can be equated with the receiver.

Reply:
What happens if you do `cx.typeck_results().expr_ty(expr);`? Doesn't that give you the resulting type of (in this case) `x.clone()`? If so, then you can compare (probably for direct equality would work) against `x`'s type (which you already have in `receiver_ty`) and emit the lint of `!=`. Could you try that out? Otherwise we can merge without that and I can build on this.

Reply:
@estebank I already tried `cx.typeck_results().expr_ty(expr);` and it returns `Self` as the type in the case of `Clone::clone` which is not equal to the receiver type unfortunately. If you have any additional ideas, I'd be happy to try them!

Reply:
Can you try calling `expr_ty_adjusted` (for both the receiver and expr)?